tests/unit/test_download.py

```diff
@@ -339,3 +340,98 @@ def test_parse_credentials():
     # URL-encoded reserved characters:
     assert auth.parse_credentials("user%3Aname:%23%40%5E@example.com") \
         == ("user:name", "#@^")
+
+
+def test_get_credentials():
+    auth = MultiDomainBasicAuth()
+    get = functools.partial(
+        auth._get_url_and_credentials,
+        allow_netrc=False,
+        allow_keyring=False
+    )
+
+    # Check URL parsing
+    assert get("http://foo:bar@example.com/path") \
+        == ('http://example.com/path', 'foo', 'bar')
+    assert auth.passwords['example.com'] == ('foo', 'bar')
+
+    auth.passwords['example.com'] = ('user', 'pass')
+    assert get("http://foo:bar@example.com/path") \
+        == ('http://example.com/path', 'user', 'pass')
+
+
+def test_get_credentials_keyring():
+    try:
+        import keyring
+        import keyring.backend
+        import keyring.core
+    except ImportError:
+        pytest.skip("keyring required for test")
+
+    auth = MultiDomainBasicAuth()
+    get = functools.partial(
+        auth._get_url_and_credentials,
+        allow_netrc=False,
+        allow_keyring=True
+    )
+
+    # Specify a backend with known responses for specific systems.
+    # Validates that we (a) pass through the correct system, and
+    # (b) pass back the correct result.
+    class GetPasswordBackendV1(object):
+        def get_password(self, system, username):
+            if system == "http://example.com/path1":
+                return (username or "") + "!password1"
+            return None
+
+    class GetPasswordBackendV2(GetPasswordBackendV1):
+        def get_credential(self, system, username):
+            if system == "http://example.com/path2":
+                return keyring.backend.Credential("username", "password2")
+            return None
+
+    keyring.backend.KeyringBackend.register(GetPasswordBackendV1)
+
+    test_keyring1 = GetPasswordBackendV1()
+    test_keyring2 = GetPasswordBackendV2()
+    orig_keyring = keyring.core.get_keyring()
+    try:
+        # Enable the old-style keyring backend. This does *not*
+        # support adding or overriding the username.
+        keyring.core.set_keyring(test_keyring1)
+
+        auth.passwords.pop('example.com', None)
+        assert get("http://example.com/path1") \
+            == ('http://example.com/path1', None, None)
+        auth.passwords.pop('example.com', None)
```

Comment:
Do you think this series of assertions could benefit from using `pytest.mark.parametrize()`? See the following code and nearby for some examples:
https://github.com/pypa/pip/blob/ebfa66361e92b3cac7c259956effcd55617c1486/tests/unit/test_utils.py#L655-L671

Reply:
I can definitely make this read better, but I think there's enough setup/teardown that it's not worth doing it multiple times (perhaps pytest has a way to set this up? But I don't know what it is and nothing in the docs are jumping out).

Reply:
You can use a module specific fixture, or use a class with a setUp function.